identity-aware-proxy/IAPClient/IAPClient.cs

```diff
@@ -48,44 +50,32 @@ class IAPClient
             string credentialsFilePath, string uri)
         {
             // Read credentials from the credentials .json file.
-            Credentials credentials;
+            ServiceAccountCredential saCredential;
             using (var fs = new FileStream(credentialsFilePath,
                 FileMode.Open, FileAccess.Read))
             {
-                credentials = NewtonsoftJsonSerializer.Instance
-                    .Deserialize<Credentials>(fs);
+                saCredential = ServiceAccountCredential
+                    .FromServiceAccountData(fs);
             }
 
             // Generate a JWT signed with the service account's private key 
             // containing a special "target_audience" claim.
             var jwtBasedAccessToken =
-                CreateAccessToken(credentials.PrivateKey, iapClientId,
-                    credentials.ClientEmail);
-
-            var body = new Dictionary<string, string>
-            {
-                { "assertion", jwtBasedAccessToken },
-                { "grant_type", "urn:ietf:params:oauth:grant-type:jwt-bearer"}
-            };
+                CreateAccessToken(saCredential, iapClientId);
 
             // Request an OIDC token for the Cloud IAP-secured client ID.
-            var httpClient = new HttpClient();
-            var httpContent = new FormUrlEncodedContent(body);
-            var result = httpClient.PostAsync(GoogleAuthConsts.OidcTokenUrl,
-                httpContent).Result;
-            var responseContent = result.Content.ReadAsStringAsync().Result;
-            if (!result.IsSuccessStatusCode)
+            var req = new GoogleAssertionTokenRequest()
             {
-                throw new HttpRequestException(string.Format(
-                    CultureInfo.CurrentCulture, "{0} {1}\n{2}",
-                    (int)result.StatusCode, result.ReasonPhrase,
-                    responseContent));
-            }
-            string token = JsonConvert.DeserializeObject<IapResponse>(
-                responseContent).IdToken;
+                Assertion = jwtBasedAccessToken
+            };
+            var result = req.ExecuteAsync(saCredential.HttpClient,
+                saCredential.TokenServerUrl, CancellationToken.None,
+                saCredential.Clock).Result;
```

Comment:
I suggest we make this sample code async method and don't use .Result() but rather await on the result of the request.

Reply:
Let's discuss this more at tomorrow's meeting.